In [1]:
import tensorflow as tf

In [2]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
#print(x_train.min(), x_train.max())
#print(x_train.shape)
#print(x_test.shape)
#print(x_train[0])

In [3]:
x_train, x_test = x_train / 255.0, x_test / 255.0
#print(x_train[0])

In [4]:
print(y_train.shape)
print(y_train[0])

(60000,)
5


###  Building the Network...

In [5]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Flatten(input_shape=(28, 28)),
    tf.keras.layers.Dense(128, activation='relu'),
    tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

In [6]:
import copy #needed for copy or deepcopy
x_train_copy = copy.copy(x_train)
print(x_train_copy.shape)
print(x_train_copy[0].flatten().shape)
x_train_copy = x_train_copy.reshape(60000, 28*28)
print(x_train_copy.shape)
print(x_train.shape)

(60000, 28, 28)
(784,)
(60000, 784)
(60000, 28, 28)


In [7]:
x_train = x_train.reshape(60000, 28*28)
x_test = x_test.reshape(10000, 28*28)

In [8]:
model = tf.keras.models.Sequential([
    #tf.keras.layers.Flatten(input_shape=(28, 28))
    tf.keras.layers.Dense(10, activation='relu', input_shape=(784,)),
    #tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

In [9]:
predictions = model(x_train[:1])
#print(predictions)
predictions = predictions.numpy()
predictions

array([[-0.18408644, -0.31474018,  0.09828976,  0.00946344, -0.45190635,
        -0.2241431 ,  0.05376392,  0.03637554, -0.21060047, -0.15021642]],
      dtype=float32)

In [10]:
tf.nn.softmax(predictions).numpy()

array([[0.09375072, 0.08226828, 0.12433947, 0.11377117, 0.07172357,
        0.09006959, 0.1189246 , 0.11687457, 0.09129767, 0.09698044]],
      dtype=float32)

### Model #1

In [11]:
loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)

In [12]:
loss_fn(y_train[:1], predictions).numpy()

2.4071727

In [13]:
model = tf.keras.models.Sequential([
    #tf.keras.layers.Flatten(input_shape=(28, 28))
    tf.keras.layers.Dense(16, activation='selu', input_shape=(784,)),
    #tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

In [14]:
model.compile(optimizer='Adam',
             loss=loss_fn,
             metrics=['accuracy'])

In [15]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 1s 517us/step - loss: 0.3940 - accuracy: 0.8899
Epoch 2/5
1875/1875 [==============================] - 1s 520us/step - loss: 0.2553 - accuracy: 0.9272
Epoch 3/5
1875/1875 [==============================] - 1s 526us/step - loss: 0.2229 - accuracy: 0.9355
Epoch 4/5
1875/1875 [==============================] - 1s 520us/step - loss: 0.1986 - accuracy: 0.9430
Epoch 5/5
1875/1875 [==============================] - 1s 518us/step - loss: 0.1820 - accuracy: 0.9473


In [16]:
model.evaluate(x_test, y_test, verbose=2)

313/313 - 0s - loss: 0.1852 - accuracy: 0.9462


[0.18515436351299286, 0.9462000131607056]

### Model  #2

In [17]:
model = tf.keras.models.Sequential([
    #tf.keras.layers.Flatten(input_shape=(28, 28))
    tf.keras.layers.Dense(16, activation='relu', input_shape=(784,)),
    #tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

In [18]:
model.compile(optimizer='RMSProp',
             loss=loss_fn,
             metrics=['accuracy'])

In [19]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 1s 495us/step - loss: 0.4404 - accuracy: 0.8792
Epoch 2/5
1875/1875 [==============================] - 1s 490us/step - loss: 0.2687 - accuracy: 0.9237
Epoch 3/5
1875/1875 [==============================] - 1s 495us/step - loss: 0.2332 - accuracy: 0.9341
Epoch 4/5
1875/1875 [==============================] - 1s 492us/step - loss: 0.2109 - accuracy: 0.9408
Epoch 5/5
1875/1875 [==============================] - 1s 502us/step - loss: 0.1957 - accuracy: 0.9455


In [20]:
model.evaluate(x_test, y_test, verbose=2)

313/313 - 0s - loss: 0.1915 - accuracy: 0.9457


[0.19151876866817474, 0.9456999897956848]

### Model #3

In [21]:
model = tf.keras.models.Sequential([
    #tf.keras.layers.Flatten(input_shape=(28, 28))
    tf.keras.layers.Dense(28, activation='relu', input_shape=(784,)),
    tf.keras.layers.Dense(14, activation='relu'),
    tf.keras.layers.Dense(7, activation='relu'),
    #tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

In [22]:
model.compile(optimizer='adam',
             loss=loss_fn,
             metrics=['accuracy'])

In [23]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 1s 565us/step - loss: 0.5104 - accuracy: 0.8493
Epoch 2/5
1875/1875 [==============================] - 1s 562us/step - loss: 0.2233 - accuracy: 0.9372
Epoch 3/5
1875/1875 [==============================] - 1s 559us/step - loss: 0.1756 - accuracy: 0.9493
Epoch 4/5
1875/1875 [==============================] - 1s 561us/step - loss: 0.1487 - accuracy: 0.9572
Epoch 5/5
1875/1875 [==============================] - 1s 553us/step - loss: 0.1311 - accuracy: 0.9617


In [24]:
model.evaluate(x_test, y_test, verbose=2)

313/313 - 0s - loss: 0.1452 - accuracy: 0.9587


[0.14522098004817963, 0.9587000012397766]

### Model #4

In [25]:
model = tf.keras.models.Sequential([
    #tf.keras.layers.Flatten(input_shape=(28, 28))
    tf.keras.layers.Dense(784, activation='tanh', input_shape=(784,)),
    #tf.keras.layers.Dense(14, activation='relu'),
    #tf.keras.layers.Dense(14, activation='softmax'),
    #tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

In [26]:
model.compile(optimizer='ftrl',
             loss=loss_fn,
             metrics=['accuracy'])

In [27]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 5s 3ms/step - loss: 2.3000 - accuracy: 0.1567
Epoch 2/5
1875/1875 [==============================] - 5s 3ms/step - loss: 2.0634 - accuracy: 0.3059
Epoch 3/5
1875/1875 [==============================] - 7s 4ms/step - loss: 1.2978 - accuracy: 0.5935
Epoch 4/5
1875/1875 [==============================] - 7s 3ms/step - loss: 0.8702 - accuracy: 0.7520
Epoch 5/5
1875/1875 [==============================] - 7s 4ms/step - loss: 0.6913 - accuracy: 0.8049


In [28]:
model.evaluate(x_test, y_test, verbose=2)

313/313 - 0s - loss: 0.6152 - accuracy: 0.8272


[0.6151817440986633, 0.8271999955177307]

### Model # 5

In [29]:
model = tf.keras.models.Sequential([
    #tf.keras.layers.Flatten(input_shape=(28, 28))
    tf.keras.layers.Dense(28, activation='sigmoid', input_shape=(784,)),
    #tf.keras.layers.Dense(14, activation=''),
    #tf.keras.layers.Dropout(0.2),
    tf.keras.layers.Dense(10)
])

In [30]:
model.compile(optimizer='adam',
             loss=loss_fn,
             metrics=['accuracy'])

In [31]:
model.fit(x_train, y_train, epochs=5)

Epoch 1/5
1875/1875 [==============================] - 1s 527us/step - loss: 0.6200 - accuracy: 0.8566
Epoch 2/5
1875/1875 [==============================] - 1s 549us/step - loss: 0.2717 - accuracy: 0.9257
Epoch 3/5
1875/1875 [==============================] - 1s 529us/step - loss: 0.2180 - accuracy: 0.9390
Epoch 4/5
1875/1875 [==============================] - 1s 548us/step - loss: 0.1878 - accuracy: 0.9473
Epoch 5/5
1875/1875 [==============================] - 1s 537us/step - loss: 0.1678 - accuracy: 0.9523


In [32]:
model.evaluate(x_test, y_test, verbose=2)

313/313 - 0s - loss: 0.1707 - accuracy: 0.9475


[0.1707497090101242, 0.9474999904632568]

### The Answer to Homework Question #1 
<p>1. Based on this article about the sigmoid vs softmax functions, why do we not want to use sigmoid in our mnist example? <br>
https://glassboxmedicine.com/2019/05/26/classification-sigmoid-vs-softmax/<br><br>
The sigmoid function treats all of the outputs individually.  The softmax algorithm treats them as interrelated values dependent on one another.  Each of the softmax values collectively sum to one.  So if one value increases another decreases. The softmax algorithm predicts the one and only one answer.  A sigmoid function can detect many different outcomes. A softmax function only predicts one.  For our MNIST example we only want to predict one outcome.  We don’t want an algorithm that will return a set of possible solutions.  Each number must be clearly identified.   We do not want an algorithm that will produce an ambiguous solution. <p>


### Results

<p>Model #1<br>
options: relu, adam<br> 
train time: 1 second<br> 
train acc: 0.9471<br>
 test acc0.9411<br>
To get best results, you<br>
I did not alter anything for this test<br>

Model #2<br>
options: relu, RMSProp<br> 
train time: 1 second<br> 
train acc: 0.9433<br>
 test acc 0.9442<br>
To get best results, you<br>
Changed optimizer to RMSProl<br>

Model #3<br>
options: relu, adam<br> 
train time: 1 second<br> 
train acc: 0.96193<br>
 test acc 0.9597<br>
To get best results, you<br>
Defined three layers</p>

Model #4<br>
options: tanh, ftrl<br> 
train time: 1 second<br> 
train acc: 0.8033<br>
 test acc 0.8244<br>
To get best results, you<br>
Changed optimizer to ftrl<br>

Model #5<br>
options: sigmoid, adam<br> 
train time: 1 second<br> 
train acc: 0.9523<br>
 test acc 0.9475<br>
To get best results, you<br>
used sigmoid activation function</p>